test/extensions/tracers/opentelemetry/resource_detectors/resource_provider_test.cc

```diff
@@ -0,0 +1,403 @@
+
+  // Simulating having a problem when creating the resource detector
+  EXPECT_CALL(factory, createResourceDetector(_, _)).WillOnce(Return(nullptr));
+
+  const std::string yaml_string = R"EOF(
+    grpc_service:
+      envoy_grpc:
+        cluster_name: fake-cluster
+      timeout: 0.250s
+    service_name: my-service
+    resource_detectors:
+      - name: envoy.tracers.opentelemetry.resource_detectors.a
+        typed_config:
+          "@type": type.googleapis.com/google.protobuf.Struct
+    )EOF";
+
+  envoy::config::trace::v3::OpenTelemetryConfig opentelemetry_config;
+  TestUtility::loadFromYaml(yaml_string, opentelemetry_config);
+
+  ResourceProviderImpl resource_provider;
+  EXPECT_THROW_WITH_MESSAGE(resource_provider.getResource(opentelemetry_config, context_),
+                            EnvoyException,
+                            "Resource detector could not be created: "
+                            "'envoy.tracers.opentelemetry.resource_detectors.a'");
+}
+
+TEST_F(ResourceProviderTest, SchemaUrl) {
+  // old resource schema is empty but updating is not. should keep updating schema
+  {
+    std::string expected_schema_url = "my.schema/v1";
+    Resource old_resource = resource_a_;
+
+    Resource updating_resource = resource_b_;
+    updating_resource.schemaUrl_ = expected_schema_url;
+
+    auto detector_a = std::make_shared<NiceMock<SampleDetector>>();
+    EXPECT_CALL(*detector_a, detect()).WillOnce(Return(old_resource));
+
+    auto detector_b = std::make_shared<NiceMock<SampleDetector>>();
+    EXPECT_CALL(*detector_b, detect()).WillOnce(Return(updating_resource));
+
+    DetectorFactoryA factory_a;
+    Registry::InjectFactory<ResourceDetectorFactory> factory_a_registration(factory_a);
+
+    DetectorFactoryB factory_b;
+    Registry::InjectFactory<ResourceDetectorFactory> factory_b_registration(factory_b);
+
+    EXPECT_CALL(factory_a, createResourceDetector(_, _)).WillOnce(Return(detector_a));
+    EXPECT_CALL(factory_b, createResourceDetector(_, _)).WillOnce(Return(detector_b));
+
+    const std::string yaml_string = R"EOF(
+    grpc_service:
+      envoy_grpc:
+        cluster_name: fake-cluster
+      timeout: 0.250s
+    service_name: my-service
+    resource_detectors:
+      - name: envoy.tracers.opentelemetry.resource_detectors.a
+        typed_config:
+          "@type": type.googleapis.com/google.protobuf.Struct
+      - name: envoy.tracers.opentelemetry.resource_detectors.b
+        typed_config:
+          "@type": type.googleapis.com/google.protobuf.StringValue
+    )EOF";
+    envoy::config::trace::v3::OpenTelemetryConfig opentelemetry_config;
+    TestUtility::loadFromYaml(yaml_string, opentelemetry_config);
+
+    ResourceProviderImpl resource_provider;
+    Resource resource = resource_provider.getResource(opentelemetry_config, context_);
+
+    EXPECT_EQ(expected_schema_url, resource.schemaUrl_);
+  }
+  // old resource schema is not empty and updating one is. should keep old schema
```

Comment:
nit: please refactor these tests to individual tests instead of a single long one. This simplifies debugging when a test is broken.

Reply:
I wasn't sure what was the common style but I agree. Fixed in: [ca96947](https://github.com/envoyproxy/envoy/pull/29547/commits/ca96947c798cf9707d72262a091fedc078789f97)